Common/src/geometry_structure.cpp

```diff
@@ -6806,6 +6782,153 @@ void CPhysicalGeometry::CompleteCommsAll(int nSends,
+  
+  iProcessor = val_global_index/nPoint_Lin[0];
+  
+  /*--- Guard against going over size. ---*/
+  
+  if (iProcessor >= (unsigned long)size)
+    iProcessor = (unsigned long)size-1;
+  
+  /*--- Move up or down until we find the processor. ---*/
+  
+  if (val_global_index >= nPoint_Cum[iProcessor])
+    while(val_global_index >= nPoint_Cum[iProcessor+1])
+      iProcessor++;
+  else
+    while(val_global_index < nPoint_Cum[iProcessor])
+      iProcessor--;
+  
+  return iProcessor;
+  
+}
+
+void CPhysicalGeometry::SortAdjacency(CConfig *config) {
+  
+#ifdef HAVE_MPI
+#ifdef HAVE_PARMETIS
+  
+  if ((rank == MASTER_NODE) && (size > SINGLE_NODE))
+    cout << "Executing the partitioning functions." << endl;
+  
+  unsigned long iPoint, jPoint, local_size;
+  vector<unsigned long>::iterator it;
+  vector<unsigned long> adjacency_vector;
+  
+  /*--- Post process the adjacency information in order to get it into the
+   CSR format before sending the data to ParMETIS. We need to remove
+   repeats and adjust the size of the array for each local node. ---*/
+  
+  if ((rank == MASTER_NODE) && (size > SINGLE_NODE))
+    cout << "Building the graph adjacency structure." << endl;
+  
+  /*--- We can already create the array that indexes the adjacency. ---*/
+  
+  xadj_size = nPoint_Lin[rank]+1;
+  xadj      = new idx_t[xadj_size];
+  xadj[0]   = 0;
+  
+  /*--- Here, we transfer the adjacency information from a multi-dim vector
+   on a node-by-node basis into a single vector container. First, we sort
+   the entries and remove the duplicates we find for each node, then we
+   copy it into the single vect and clear memory from the multi-dim vec. ---*/
+
+  unsigned long total_adj_size = 0;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    
+    /*--- For each point, sort the adjacency in ascending order
+     so that we can remove duplicates and complete the size for
+     unique set of adjacent nodes for that point. ---*/
+    
+    sort(adj_nodes[iPoint].begin(), adj_nodes[iPoint].end());
+    it = unique(adj_nodes[iPoint].begin(), adj_nodes[iPoint].end());
+    local_size = it - adj_nodes[iPoint].begin();
+    adj_nodes[iPoint].resize(local_size);
+    total_adj_size += local_size;
+  }
+  /*--- reserve allocated space to improve push_back behavior
+  in the following loop */
+  adjacency_vector.reserve(total_adj_size);
+
+  for (iPoint = 0; iPoint < nPoint; iPoint++){
+    local_size = adj_nodes[iPoint].size();
+    /*--- Move the sorted adjacency into a 1-D vector for all
+     points in order to make creating the copy for ParMETIS easier. ---*/
+    
+    for (jPoint = 0; jPoint < local_size; jPoint++) {
+      adjacency_vector.push_back(adj_nodes[iPoint][jPoint]);
+    }
+    
+    /*--- Increment the starting index for the next point (CSR). ---*/
+    
+    xadj[iPoint+1] = xadj[iPoint] + local_size;
+    
+    /*--- Free vector memory as we go. ---*/
+    
+    vector<unsigned long>().swap(adj_nodes[iPoint]);
```

Comment:
Thanks for putting up with me @MicK7, regarding your comment about memory usage, I tested with one of the small cases in the repo and all seems fine.
I would say however that this swap does not do much for the peak memory usage, since everything is allocated at this stage and it is still up to the OS whether you see memory usage dropping by doing this or not. The swap to free the entire vector should be enough.

Reply:
Ok thank you for testing

Reply:
A comment here.. this routine is only executed once in the pre-processing, so raw performance was not as critical for me as making sure that the largest mesh possible can be processed by minimizing peak memory. So, what I was trying to do there, since we need to prepare the adjacency, was to do it one point at a time with an attempt to clear that point's memory immediately so that our peak memory might stay roughly constant during this operation (rather than going to 2x as we shuffle the adjacency).

Reply:
With the two passes, the code try to do a reserve for the adjacency_vector  allowing around 2x the memory if it possible, thus when doing a push_back on the adjacency_vector there is no new allocation in the background. If the reserve fails, it should continue working and grow the adjacency_vector as usual.

Reply:
I got the idea @economon but std vector grows exponentially when it runs out of capacity and for a while both old and new buffers need to exist to move elements, so even if on the last grow the capacity becomes just right the peak working set might be larger than reserving just enough memory at the beginning... maybe, idk.